### PR TITLE
chore(install): always report status complete on exit

### PR DESCRIPTION
**Pull request by ctrombley**
The main changes here are:
1. Use `defer` to ensure completion status is sent
2. Mutate remaining AVAILABLE recipe states to FAILED during completion.
3. Add a `RecipeSelected` hook to allow the `TerminalReporter` to display the selected integrations without causing recipe state to be modified.

There is also some renaming for clarity, notably `StatusReporter` to `StatusSubscriber` and `StatusRollup` to `InstallStatus`.

### CODE REVIEW

File: internal/install/execution/install_status.go [hunk 1]
```diff
@@ -9,32 +9,32 @@ import (
...
-	EntityGUIDs     []string `json:"entityGuids"`
-	Statuses        []Status `json:"recipes"`
-	Timestamp       int64    `json:"timestamp"`
-	LogFilePath     string   `json:"logFilePath"`
-	statusReporters []StatusReporter
+	EntityGUIDs     []string       `json:"entityGuids"`
+	Statuses        []RecipeStatus `json:"recipes"`
+	Timestamp       int64          `json:"timestamp"`
+	LogFilePath     string         `json:"logFilePath"`
+	statusReporters []StatusSubscriber
```
Comment (zlesnr) [inline internal/install/execution/install_status.go:19]: Consider rename `statusReporters` to `statusSubscribers` to match the reference type.

**Review by sanderblue — APPROVED** [review]

File: internal/install/execution/install_status.go [hunk 2]
```diff
@@ -61,126 +61,134 @@ func NewStatusRollup(reporters []StatusReporter) *StatusRollup {
...
-		if err := r.ReportRecipesAvailable(s, recipes); err != nil {
+		if err := r.RecipesAvailable(s, recipes); err != nil {
 			log.Errorf("Could not report recipe execution status: %s", err)
 		}
 	}
 }
 
-func (s *StatusRollup) ReportRecipeInstalled(event RecipeStatusEvent) {
-	s.withRecipeEvent(event, StatusTypes.INSTALLED)
+func (s *InstallStatus) RecipesSelected(recipes []types.Recipe) {
```
Comment (zlesnr) [inline internal/install/execution/install_status.go:84]: Is the method name a question here, or an action being taken?

**Review by zlesnr — DISMISSED** [review]

File: internal/install/execution/install_status.go (hunk 2 above)
Comment (ctrombley) [reply internal/install/execution/install_status.go:84]: This looks like the diff tool struggling.  `RecipesSelected` has been added, and `ReportRecipeInstalled` has also been renamed to `RecipeInstalled`.

File: internal/install/execution/install_status.go (hunk 1 above)
Comment (ctrombley) [reply internal/install/execution/install_status.go:19]: 👍

File: internal/install/recipe_installer.go [hunk 3]
```diff
@@ -101,7 +101,7 @@ func (i *RecipeInstaller) Install() error {
 	// Execute the discovery process, exiting on failure.
 	m, err := i.discover()
 	if err != nil {
-		return i.fail(err)
+		return err
```
Comment (zlesnr) [inline internal/install/recipe_installer.go:104]: 👍🏼

File: internal/install/execution/install_status.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+	for _, r := range s.statusSubscriber {
+		if err := r.RecipeSkipped(s, event); err != nil {
+			log.Errorf("Error writing recipe status for recipe %s: %s", event.Recipe.Name, err)
+		}
+	}
+}
+
+func (s *InstallStatus) InstallComplete() {
+	s.completed()
```
Comment (zlesnr) [inline internal/install/execution/install_status.go:148]: 👍🏼

**Review by zlesnr — APPROVED** [review]

**Pull request merged**